developer documentation for placeholders

elee1766 opened the pull request:
also adds a docker command to the readme that might help people working with this repo

i run caddy already locally, so it's a bit of a pain to have to disable admin. (also i cant bind to 80/443 without root)



hosting it here
https://caddy-website.put.gay/docs/extending-caddy/placeholders

elee1766 committed:
noot

elee1766 committed:
noot

elee1766 committed:
formatting

elee1766 committed:
links

elee1766 committed:
links

elee1766 committed:
better explained

elee1766 committed:
grammar

elee1766 committed:
grammar

elee1766 committed:
fix ref

elee1766 committed:
more explained

mholt commented on src/docs/markdown/extending-caddy/placeholders.md:
```diff
@@ -0,0 +1,138 @@
+
+# Placeholders
+
+In Caddy, placeholders are processed by each individual plugin themselves. They are not parsed at config time, but instead preserved and processed at runtime.
```
> They are not parsed at config time, but instead preserved and processed at runtime.

Technically, this depends if they're evaluated during Provision or later on. Maybe more specfic:

```suggestion
In Caddy, placeholders are processed by each individual plugin as needed; they do not automatically work everywhere.
```

mholt commented on src/docs/markdown/extending-caddy/placeholders.md:
```diff
@@ -0,0 +1,138 @@
+
+## Placeholder Internals
+
+Internally, placeholders are simply a string in format `{foo.bar}` used as valid configuration values, which is later parsed at runtime.
```
```suggestion
Placeholders are a string in the format `{foo.bar}` used as dynamic configuration values, which is later evaluated at runtime.
```

mholt commented on src/docs/markdown/extending-caddy/placeholders.md:
```diff
@@ -0,0 +1,138 @@
+
+Placeholders-like strings which start with a dollar sign (`{$FOO}`), are evaulated at Caddyfile parse time, and do not need to be dealt with by your plugin. This is because these are not placeholders, but Caddyfile-specific [environmental variable substitution](/docs/caddyfile/concepts#environmental-variables), they just happen to share the `{}` syntax.
+
+It is therefore important to understand that `{env.HOST}` is inherently different from something like `{$HOST}`
```
```suggestion
It is therefore important to understand that `{env.HOST}` is inherently different from something like `{$HOST}`.
```

elee1766 committed:
noot

elee1766 committed:
noot

mholt commented on src/docs/markdown/extending-caddy/placeholders.md:
```diff
@@ -0,0 +1,140 @@
+
+It is therefore important to understand that `{env.HOST}` is inherently different from something like `{$HOST}`
+
+As an example, see the following caddyfile:
```
```suggestion
As an example, see the following Caddyfile:
```

mholt commented on src/docs/markdown/extending-caddy/placeholders.md:
```diff
@@ -0,0 +1,140 @@
+
+## How to use placeholders in your plugin
+
+#### Parse the raw placeholder value in your unmarshaler
```
I don't really think this section adds any value / says anything interesting. It can probably be removed...

elee1766 replied:
hmmm, my thought is that some people may be confused thinking they might need to use a special function to parse a placeholder.

if you don't think thats a problem, we can gut it

mholt replied:
What if we remove the section but put what you're conveying in the parent section ("Implementing support for placeholders" as a heading I suggest):

> Placeholders are evaluated at runtime (not at Caddyfile parsing or config loading). When you are unmarshaling a Caddyfile, do not evaluate placeholders at that time. Placeholders may be evaluated during `Provision()` or during your module's execution.

mholt commented on src/docs/markdown/extending-caddy/placeholders.md:
````diff
@@ -0,0 +1,140 @@
+}
+```
+
+#### Evaluate the placeholder during Match or Serve
````
I would reframe this section and the next to be more general.

It boils down to whether a Replacer is already available or not. If one has already been made, as is the case with the HTTP handler chain, it can be obtained from the context. I think we can talk about it with that as the example, but less specific to "Match" or "ServeHTTP". The only real difference is how you get the Replacer: either from a context, or make a new one.

elee1766 replied:
do you think it should be reframed to be more general, or there be instead something two articles/examples showing

1. explaining how to grab a new replacer, and using that to resolve a placeholder
2. explaining how to create a replacer from a context.Context.

with a blurb saying that to use placeholder in code, pass the string into a replacer?

mholt replied:
I think this one article is enough; mainly let's show examples of both in this one.

So show how to get it from the context, where it will often be stored, and also how to create a new one if there isn't one already available.

mholt reviewed (changes requested):
Thanks for taking the initiative on this! I think this will be a great addition to the docs. I have some thoughts for your review.

elee1766 committed:
Update src/docs/markdown/extending-caddy/placeholders.md

Co-authored-by: Matt Holt <mholt@users.noreply.github.com>

elee1766 committed:
Update src/docs/markdown/extending-caddy/placeholders.md

Co-authored-by: Matt Holt <mholt@users.noreply.github.com>

elee1766 committed:
Update src/docs/markdown/extending-caddy/placeholders.md

Co-authored-by: Matt Holt <mholt@users.noreply.github.com>

elee1766 committed:
Update src/docs/markdown/extending-caddy/placeholders.md

Co-authored-by: Matt Holt <mholt@users.noreply.github.com>

elee1766 commented:
@mholt accepted all the grammar changes, had two questions for you on your last two comments that will change what i write to replace the last section. i agree that it's weak.

mholt commented on src/docs/markdown/extending-caddy/placeholders.md:
```diff
@@ -0,0 +1,140 @@
+
+If you are not yet familiar with placeholders, start by reading [here](/docs/conventions#placeholders)!
+
+## Placeholder Internals
```
```suggestion
## Placeholder Overview
```

mholt commented on src/docs/markdown/extending-caddy/placeholders.md:
```diff
@@ -0,0 +1,140 @@
+Some users may immediately notice that this means it is impossible to use the `{$ENV}` syntax in a JSON config. The solution to this is to process such placeholders at Provision time, which is covered below.
+
+
+## How to use placeholders in your plugin
```
```suggestion
## Implementing support for placeholders
```

elee1766 replied:
i chose `Implementing placeholder support` to better match the article title `implementing placeholders`

mholt reviewed (changes requested):
Thanks for the updates! Almost there. I hope I'm able to answer your questions alright. They are good questions, thanks for working on this!

elee1766 committed:
noot

elee1766 committed:
noot

elee1766 commented:
@mholt i rewrote the last section into a single examples section with an intro - it feels a lot nicer. 

also rewrote it to be more instructional/imperative, as at this point, the user should have already read the overview, and just probably wants to know what to do.

only thought - should i also include an example of parsing a string? or should people be expected to know that.

elee1766 committed:
noot

elee1766 committed:
noot

elee1766 committed:
noot

francislavoie commented on src/docs/markdown/extending-caddy/placeholders.md:
```diff
@@ -82,56 +82,30 @@ Since `srv1` used `{env.HOST}`, a normal placeholder, it was parsed as its own r
-#### Alternatively, resolve the placeholder during Provision
-
-If you only use global placeholders, like `env`, then you may initialize a global replacer at provision time, and use it to replace such values. This also allows users of config file formats other than Caddyfile to use environmental variables.
+In this example, we are using a newly constructed replacer to process placeholders. It has access to placeholders such as `{env.HOST}`, but NOT `{http.request.uri}`
```
```suggestion
In this example, we are using a newly constructed replacer to process placeholders. It has access to [global placeholders](/docs/conventions#placeholders) such as `{env.HOST}`, but NOT request placeholder such as `{http.request.uri}`
```

elee1766 replied:
i accepted this manually.

elee1766 committed:
noot

francislavoie commented on src/docs/markdown/extending-caddy/placeholders.md:
```diff
@@ -0,0 +1,113 @@
+
+## Implementing placeholder support
+
+You should not process placeholders when ummarshaling your Caddyfile. Instead, unmarshal the placeholders as strings in your configuration and evaluate them during either your module's execution or `Provision()` using a `caddy.Replacer`.
```
```suggestion
You should not process placeholders when ummarshaling your Caddyfile. Instead, unmarshal the placeholders as strings in your configuration and evaluate them during either your module's execution (e.g. `ServeHTTP()` for HTTP handlers, `Match()` for matchers, etc.) or in the `Provision()` step, using a `caddy.Replacer`.
```

elee1766 replied:
reads well to me

francislavoie reviewed (approved):
After that last thing :point_up:, looks good to me, can merge.

elee1766 committed:
Update src/docs/markdown/extending-caddy/placeholders.md

Co-authored-by: Francis Lavoie <lavofr@gmail.com>

elee1766 commented:
done

mholt reviewed (approved):
Great! I think this looks better now. :) Thank you!
